core/core.go

```diff
@@ -62,20 +62,42 @@ type baseSchema struct {
 }
 
 func (r *baseSchema) Solve(ctx context.Context, st llb.State, marshalOpts ...llb.ConstraintsOpt) (*filesystem.Filesystem, error) {
-	def, err := st.Marshal(ctx, append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)...)
+	marshalOpts = append([]llb.ConstraintsOpt{llb.Platform(r.platform)}, marshalOpts...)
+
+	inputCfg, err := filesystem.ImageConfigFromState(ctx, st)
+	if err != nil {
+		return nil, err
+	}
+
+	input, err := st.Marshal(ctx, marshalOpts...)
 	if err != nil {
 		return nil, err
 	}
-	_, err = r.gw.Solve(ctx, bkgw.SolveRequest{
+
+	res, err := r.gw.Solve(ctx, bkgw.SolveRequest{
 		Evaluate:   true,
-		Definition: def.ToPB(),
+		Definition: input.ToPB(),
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	// FIXME: should we create a filesystem from `res.SingleRef()`?
-	return filesystem.FromDefinition(def), nil
+	bkref, err := res.SingleRef()
+	if err != nil {
+		return nil, err
+	}
+
+	resSt, err := bkref.ToState()
+	if err != nil {
+		return nil, err
+	}
+
+	img, err := filesystem.ImageFromStateAndConfig(ctx, resSt, inputCfg, marshalOpts...)
```

Comment:
I'm not sure if this actually needs `marshalOpts` since we already used it above and now we're just marshaling the result. :thinking: